### PR TITLE
Adding PDU.SET implementation to MOGroup.

### PR DESCRIPTION
**Pull request by mario-haag**

### CODE REVIEW

File: snmpman/src/test/java/com/oneandone/snmpman/AbstractSnmpmanTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        target.setCommunity(new OctetString(community));
+        target.setAddress(targetAddress);
+        target.setRetries(2);
+        target.setTimeout(1500);
+        target.setVersion(SnmpConstants.version2c);
+        return target;
+    }
+
+    public static List<TableEvent> getResponse(final OID query, int port) throws Exception {
+        return getResponse(query, port, "public");
```
Comment (sfuhrm) [inline snmpman/src/test/java/com/oneandone/snmpman/AbstractSnmpmanTest.java]: This seems to be the COMMUNITY from SnmpmanSetTest? I recommend to put it to AbstractSnmpTest?

File: snmpman/src/test/java/com/oneandone/snmpman/AbstractSnmpmanTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    void assertThatOidHasValue(OID oid, String expectedValue) throws Exception {
+        List<TableEvent> responses1 = getResponse(oid, 10009);
```
Comment (sfuhrm) [inline snmpman/src/test/java/com/oneandone/snmpman/AbstractSnmpmanTest.java]: 10009 is a magic number
Make it a constant somewhere :)

File: snmpman/src/test/java/com/oneandone/snmpman/AbstractSnmpmanTest.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.testng.Assert.assertTrue;
+
+public class AbstractSnmpmanTest {
```
Comment (sfuhrm) [inline snmpman/src/test/java/com/oneandone/snmpman/AbstractSnmpmanTest.java]: Maybe SnmpmanIntegrationTest could also derive from AbstractSnmpmanTest?

File: snmpman/src/test/java/com/oneandone/snmpman/SnmpmanSetTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                new VariableBinding(new OID(".1.3.6.1.2.1.31.1.1.1.15"), new Gauge32(999)),
+                new VariableBinding(OID_UNCOVERED_FROM_SCOPE, new OctetString("not covered scope"))};
+
+        ResponseEvent responseEvent = getResponseEvent(OID_TABLE, OID_ROW_INDEX, bindings);
+        assertEquals(responseEvent.getResponse().getErrorStatus(), SnmpConstants.SNMP_ERROR_NO_CREATION);
+        assertThatOidHasValue(new OID(".1.3.6.1.2.1.31.1.1.1.1"), "Gi0/1");
+        assertThatOidHasValue(new OID(".1.3.6.1.2.1.31.1.1.1.15"), "1000");
+    }
+
+    ResponseEvent getResponseEvent(OID rowStatusColumnOID, OID rowIndex, VariableBinding[] bindings) throws IOException {
```
Comment (sfuhrm) [inline snmpman/src/test/java/com/oneandone/snmpman/SnmpmanSetTest.java]: better method name: "setOidValue()"?

File: snmpman/src/main/java/com/oneandone/snmpman/snmp/MOGroup.java [hunk 5]
```diff
@@ -130,19 +146,61 @@ public boolean next(final SubRequest request) {
...
     }
 
+    /**
+     * If any ErrorStatus, except the implicit PDU.noError, has occurred during commit then the old values are written.
+     * @param request The SubRequest to handle.
+     */
     @Override
     public void undo(final SubRequest request) {
-        request.setErrorStatus(PDU.undoFailed);
+        tmpVariableBindings.forEach(variableBindings::put);
```
[the remark this answers is not in the document]
Comment (mario-haag) [reply snmpman/src/main/java/com/oneandone/snmpman/snmp/MOGroup.java]: Was refactored with Optional<Variable> approach.

File: snmpman/src/main/java/com/oneandone/snmpman/snmp/MOGroup.java [hunk 6]
```diff
@@ -130,19 +146,61 @@ public boolean next(final SubRequest request) {
...
+        request.getStatus().setPhaseComplete(true);
     }
 
+    /**
+     * If any ErrorStatus, except the implicit PDU.noError, has occurred during commit then the old values are written.
+     * @param request The SubRequest to handle.
+     */
     @Override
     public void undo(final SubRequest request) {
-        request.setErrorStatus(PDU.undoFailed);
```
[the remark this answers is not in the document]
Comment (mario-haag) [reply snmpman/src/main/java/com/oneandone/snmpman/snmp/MOGroup.java]: Now the scenario is covered due inconsistent values during commit.

File: snmpman/src/main/java/com/oneandone/snmpman/snmp/MOGroup.java [hunk 7]
```diff
@@ -130,19 +134,67 @@ public boolean next(final SubRequest request) {
...
+        if (request.getIndex() > 0) {
+            RequestStatus status = request.getStatus();
+            if ((request.getUndoValue() != null) &&
+                    (request.getUndoValue() instanceof Variable)) {
+                variableBindings.put(request.getVariableBinding().getOid(), (Variable) request.getUndoValue());
+                status.setPhaseComplete(true);
+            } else {
+                status.setErrorStatus(SnmpConstants.SNMP_ERROR_UNDO_FAILED);
```
Comment (sfuhrm) [inline snmpman/src/main/java/com/oneandone/snmpman/snmp/MOGroup.java]: if `getUndoValue()` returns `null`, then the undo operation is not failed but has to be performed by **removing** the entry in the `variableBindings`

**Pull request merged**